pkg/api/v1/clients/memory/resource_client_test.go

```diff
@@ -62,6 +62,178 @@ var _ = Describe("Base", func() {
 		Expect(listret[0]).NotTo(BeIdenticalTo(listret2[0]))
 	})
 
+	Context("listing resources", func() {
+		var (
+			obj1 *v1.MockResource
+			obj2 *v1.MockResource
+			obj3 *v1.MockResource
+			obj4 *v1.MockResource
+			obj5 *v1.MockResource
+		)
+
+		BeforeEach(func() {
+			obj1 = &v1.MockResource{
+				Metadata: &core.Metadata{
+					Name:      "name1",
+					Namespace: "ns1",
+					Labels: map[string]string{
+						"key": "val1",
+					},
+				},
+			}
+			obj2 = &v1.MockResource{
+				Metadata: &core.Metadata{
+					Name:      "name2",
+					Namespace: "ns2",
+					Labels: map[string]string{
+						"key": "val2",
+					},
+				},
+			}
+			obj3 = &v1.MockResource{
+				Metadata: &core.Metadata{
+					Name:      "name3",
+					Namespace: "ns1",
+					Labels: map[string]string{
+						"key": "val2",
+					},
+				},
+			}
+			obj4 = &v1.MockResource{
+				Metadata: &core.Metadata{
+					Name:      "name4",
+					Namespace: "ns2",
+					Labels: map[string]string{
+						"key": "val3",
+					},
+				},
+			}
+			obj5 = &v1.MockResource{
+				Metadata: &core.Metadata{
+					Name:      "name5",
+					Namespace: "ns1",
+					Labels: map[string]string{
+						"key": "val3",
+					},
+				},
+			}
+
+			_, err := client.Write(obj1, clients.WriteOpts{})
+			Expect(err).NotTo(HaveOccurred())
+			_, err = client.Write(obj2, clients.WriteOpts{})
+			Expect(err).NotTo(HaveOccurred())
+			_, err = client.Write(obj3, clients.WriteOpts{})
+			Expect(err).NotTo(HaveOccurred())
+			_, err = client.Write(obj4, clients.WriteOpts{})
+			Expect(err).NotTo(HaveOccurred())
+			_, err = client.Write(obj5, clients.WriteOpts{})
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("lists all resources when empty namespace is provided", func() {
+			resources, err := client.List("", clients.ListOpts{})
+			Expect(err).NotTo(HaveOccurred())
+
+			// resources are sorted by namespace, then name
+			expectedResourceNames := []string{
+				"name1", "name3", "name5", // ns1
+				"name2", "name4", // ns2
+			}
+			Expect(resources).To(HaveLen(len(expectedResourceNames)))
+			for i, r := range resources {
+				Expect(r.GetMetadata().GetName()).To(Equal(expectedResourceNames[i]))
+			}
```

Comment:
nit: Any reason we don't do something like:
```suggestion
			Expect(resources).To(HaveLen(len(expectedResourceNames)))
			Expect(resources).To(ContainElements(obj1, obj3, obj5, obj2, obj4))
```

We could have a standalone test for proper ordering, but it feels like this way we could remove the need to refine all the names and just reference the objects directly

Reply:
do you mind if we keep as-is for now? currently, the resources returned from the List call contain extra metadata and so i can't directly use ContainElements, need to define our own matcher. it's taking me a bit to get it working correctly so maybe we can save that for a follow-up?